### DISCUSSION TITLE
Task.OnExeption()

### DISCUSSION THREAD

**Discussion by mkmkmk** [dotnet/runtime#106803] [Ideas] [answered]
I just wrote this:

`public static class TaskExt`
`{`
`    public static async Task OnException(this Task task, Action<Exception> action)`
`    {`
`        try`
`        {`
`            await task;`
`        }`
`        catch (Exception ex)`
`        {`
`            action(ex);`
`        }`
`    }`
`}`

It's mainly about handling exceptions of autonomous “set-and-forget” tasks.

The method returns a task to which a continuation can be added.
The continuation will be executed regardless of whether an exception occurred or not.

If an exception occurs, the continuation will be executed after the exception handling action.

The method also catches exceptions thrown when the task is canceled (OperationCanceledException, TaskCanceledException).

Is there something like this in .NET already?
Does it make sense?
What do you think?

`+` tests:

`[Test]`
`public void TaskOnExceptionTest()`
`{`
`    static async Task SomeTask()`
`    {`
`        await Task.Delay(100);`
`        throw new Exception("test exception");`
`    }`
`    int stage = 0;`
`    SomeTask()`
`        .OnException(ex => Assert.That(stage++, Is.Zero))`
`        .ContinueWith(prev => Assert.That(stage++, Is.EqualTo(1)))`
`        .Wait();`
`    Assert.That(stage, Is.EqualTo(2));`
`}`

`[Test]`
`public void TaskOnExceptionCancelTest()`
`{`
`    static async Task SomeTask(CancellationToken cancellationToken)`
`        => await Task.Delay(-1, cancellationToken);`
`    var cancelSource = new CancellationTokenSource();`
`    Task.Delay(1000).ContinueWith(ant => cancelSource.Cancel());`
`    int stage = 0;`
`    SomeTask(cancelSource.Token)`
`        .OnException(ex => Assert.That(stage++, Is.Zero))`
`        .ContinueWith(prev => Assert.That(stage++, Is.EqualTo(1)))`
`        .Wait();`
`    Assert.That(stage, Is.EqualTo(2));`
`}`

**Comment by huoyaoyuan** [comment, marked as answer]
Basically,

```csharp
task.ContinueWith(t =>
{
    if (t.IsFaulted)
    {
        action(t.Exception);
    }
    else if (t.IsCanceled)
    {
        action(new TaskCanceledException());
    }
})
```



> Does it make sense?

In .NET we don't actually use callback chain. Mixing `await` with callback chain just make things worse.

**Reply by mkmkmk** [reply]
ok, thanks
